### DISCUSSION TITLE
A little freeze on changing tab

### DISCUSSION THREAD

**Discussion by oolxg** [pointfreeco/swift-composable-architecture#1191] [Q&A]
Hello everybody, 

i've found some issue, makes my app freeze for a little

I have 
```
TabView(selection: viewStore.binding(\.$pageIndex))  {
 ...
 }
.tabViewStyle(.page(indexDisplayMode: .never))
```

and when i change page(tab), my app freezes for a little.

Without selection or when i use `@State` as selection, there're no freezes or similar

Also when i do the following 
```
TabView(selection: $selection) { ... }
    .tabViewStyle(.page(indexDisplayMode: .never))
     .onChange(of: selection) { let newValue in
        viewStore.send(.userChangedPage(newValue))
    }



...
case .userChangedPage(let newPageIndex):
    state.pageIndex = newPageIndex // 1
    return .none
...
```

i have the same freeze. What could be the reason for this and how could it be fixed?
(If in (1) i don't assign `newPageIndex` to `state.pageIndex`, there's no freeze)

**Comment by mbrandonw** [comment]
Can you show more code of what your view looks like, including how the `viewStore` is constructed?

**Reply by oolxg** [reply]
Yeah, sure - [boop](https://pastebin.com/1CD1XiUT)

I've found the reason of the freezes - if i create a custom `static func == (lhs: MyState, rhs: MyState) -> Bool { } `, in which i ignore comparing of `pageIndex`, there's no freeze

**Reply by mbrandonw** [reply]
Ok, glad you figured it out.

In addition to that I will say that you are observing a lot of state that you aren't actually using. This will cause views to recompute more often than needed, and could slow things down.

For example, `MangaReadingView` and `MangaReadingView` could both do:

```swift
WithViewStore(self.store.stateless) { viewStore in 
  ...
}
```

...since both views don't actually read state. They only send actions.

And then other views that do need state could probably `.scope` the store to chisel the state down to something smaller.

But also, no need to do this until you really do see slow downs. Just a tip.

**Reply by oolxg** [reply]
Thank you for the tip, but the problem isn't solved completely

I need `currentPage` in `MangaReadingView `, but if i don't include it in `static func ==(lhs: Self, rhs: Self)`, the view will not be updated (line 108-111)

**Reply by mbrandonw** [reply]
Is there a reason you are providing custom `==` implementations? It would be best to use the default synthesized conformance so that you can make sure all fields are being accounted for.

**Reply by oolxg** [reply]
It doesn't matter if i provide `==` or not - the lag remains

I mentioned `==` just to show, that if i don't compare `pageIndex`'s, the lag disappears

**Reply by mbrandonw** [reply]
It's hard to know what is going on without code I can run myself, but I would suggest reducing the amount of state you are observing like I mentioned before. Many of your view stores can be stateless, and the others can probably be significantly reduced.
